pkg/leaderelection/candidate.go

```diff
@@ -43,94 +50,106 @@ type Candidate struct {
 
 	// registry being used for leader election
 	registry registry.IRegistry
+
+	errCh chan error
 }
 
 // Run starts the leader election loop. Run will not return
 // before leader election loop is stopped by ctx or it has
 // stopped holding the leader lease
 func (c *Candidate) Run(ctx context.Context) error {
-	logger.Ctx(ctx).Info("attempting to acquire leader lease")
-
-	leadCtx, leadCancel := context.WithCancel(ctx)
-	// outer wait loop runs when a instance has successfully acquired lease
-	wait.Until(func() {
-		retryCtx, retryCancel := context.WithTimeout(leadCtx, c.config.RenewDeadline)
-		defer retryCancel()
-
-		// inner wait retry loop for faster retries if failed to acquire lease
-		acquired := false
-		wait.JitterUntil(func() {
-			acquired = c.tryAcquireOrRenew(retryCtx)
-
-			if !acquired {
-				logger.Ctx(retryCtx).Infow("failed to acquire lease, retrying...", "nodeID", c.nodeID)
-				return
-			}
-
-			// if succeeded, we can break the inner loop by cancelling context and renew in outer loop
-			logger.Ctx(retryCtx).Infow("successfully acquired lease", "nodeID", c.nodeID)
-			retryCancel()
-		}, c.config.RetryPeriod, JitterFactor, true, retryCtx.Done())
-
-		// OnStartedLeading if new leader
-		if acquired && !c.leader {
-			c.leader = true
-			go func() {
-				err := c.config.Callbacks.OnStartedLeading(leadCtx)
-				if err != nil {
-					logger.Ctx(leadCtx).Errorf("on started leading callback returned with err : %s", err.Error())
-					leadCancel()
-				}
-			}()
-		}
-	}, c.config.RenewDeadline, leadCtx.Done())
+	var err error
 
-	// context returned done, release the lease
-	logger.Ctx(leadCtx).Info("context done, releasing lease")
-	c.release(leadCtx)
-	return leadCtx.Err()
-}
+	logger.Ctx(ctx).Info("registering node with registry")
+	c.nodeID, err = c.registry.Register(c.config.Name, c.config.LeaseDuration)
```

Comment:
This register is required because once the node goes away, the session would become invalid and all nodes it created in consul would be deleted? Is that understanding correct?

Reply:
should it be on config name or id then?

Reply:
Yes, Assuming by node you mean keys created and acquired under registry.
in our case we are acquiring 2 
/registry/nodes/<id>
/leader/election

Reply:
>should it be on config name or id then?
On register, consul issues a session id, we can pass both name and id, for now passing id is optional as we dont have any use case, can be added later if required